### PR TITLE
Implement `cd -` to return to the previous directory

### PR DESCRIPTION
**Pull request by twe4ked**
Fixes https://github.com/nushell/nushell/issues/351.

### CODE REVIEW

**Comment by jonathandturner** [comment]
Looks good. Can you add a test so we can make sure we don't break it in the future by accident?

**Comment by twe4ked** [comment]
Force pushed to fix a conflict and add tests. 🤞 they work with windows paths.

**Comment by wycats** [comment]
Why are the tests failing?

**Comment by androbtech** [comment]
Windows treats (adds or not double quotes) to strings read from nu's output and that's what causes it. Other than that, `cd -` works as expected.

**Comment by twe4ked** [comment]
This is the failure:

```
 left: `"/d/a/1/s/tests/fixtures/nuplayground/cd_back_test/odin/d/a/1/s/tests/fixtures/nuplayground/cd_back_test/andres/d/a/1/s/tests/fixtures/nuplayground/cd_back_test/odin"`,
right: `"D:\\a\\1\\s/tests/fixtures/nuplayground/cd_back_test/odin|D:\\a\\1\\s/tests/fixtures/nuplayground/cd_back_test/andres|D:\\a\\1\\s/tests/fixtures/nuplayground/cd_back_test/odin"`'
```

I've been playing with a fix that might work locally, it's still a bit WIP:

```rust
let output: Vec<_> = output
    .split('|')
    .map(|p| p.replace(std::env::current_dir().unwrap().to_str().unwrap(), ""))
    .collect();

let odin_dir = format!("/{}/{}", full_path, "odin");
let andres_dir = format!("/{}/{}", full_path, "andres");
assert_eq!(output, vec![odin_dir.clone(), andres_dir, odin_dir]);
```

**Comment by jonathandturner** [comment]
Would it make sense to take the path, read it into a Path (or PathBuf), and then only check the last part of the directory (Rather than checking the whole thing)?

**Comment by twe4ked** [comment]
That's a good idea, I just thought of a bug though–I don't think we can `cd` into a directory called `-`. I'll definitely get back to this but I'm pretty flat out this week.

**Comment by twe4ked** [comment]
“Fix tests on Windows” 🤞

File: src/shell/filesystem_shell.rs [hunk 1]
```diff
@@ -177,23 +183,39 @@ impl Shell for FilesystemShell {
...
+                                PathBuf::from(self.last_path.clone()).as_path(),
+                            ) {
+                                Ok(p) => p,
+                                Err(_) => {
+                                    return err;
+                                }
+                            }
+                        } else {
+                            return err;
+                        }
```
Comment (twe4ked) [inline src/shell/filesystem_shell.rs]: Not very happy with the control flow here but I'm not seeing an easy way to make it clearer.

**Comment by twe4ked** [comment]
> Would it make sense to take the path, read it into a Path (or PathBuf), and then only check the last part of the directory (Rather than checking the whole thing)?

Wait… you were meaning in the tests. I changed the actual code (while fixing the bug I mentioned). I'll give that a go in the tests.

File: src/shell/filesystem_shell.rs [hunk 2]
```diff
@@ -177,23 +183,39 @@ impl Shell for FilesystemShell {
...
                     Ok(p) => p,
                     Err(_) => {
-                        return Err(ShellError::labeled_error(
+                        let err = Err(ShellError::labeled_error(
                             "Can not change to directory",
                             "directory not found",
                             v.span().clone(),
                         ));
+
+                        if path.ends_with("-") {
```
Comment (jonathandturner) [inline src/shell/filesystem_shell.rs]: Is it any path that ends with "-" or if the whole thing is "-"?  I'm not super familiar with the command

**Comment by jonathandturner** [comment]
@twe4ked  - yeah exactly. Sorry, I could have said that a bit better

File: src/shell/filesystem_shell.rs (hunk 2 above)
Comment (twe4ked) [reply src/shell/filesystem_shell.rs]: Yeah, [the docs](https://doc.rust-lang.org/std/path/struct.Path.html#method.ends_with) are a bit ambiguous but it matches the whole path component:

```
error: Can not change to directory
- shell:1:15
1 |             cd foo-
  |                ^^^^ directory not found
```

**Comment by twe4ked** [comment]
Now I'm confused. Might try get a development environment set up on my PC.

```
running 2 tests
test cd_directory_not_found ... ok
The syntax of the command is incorrect.
The syntax of the command is incorrect.
The syntax of the command is incorrect.
test cd_back ... FAILED
```

File: src/shell/filesystem_shell.rs (hunk 2 above)
Comment (vivainio) [reply src/shell/filesystem_shell.rs]: How about "cd foo/-"? Looks like that would match with ends_with

**Comment by twe4ked** [comment]
This will need to incorporate `Playground` changes from `master`.

https://github.com/nushell/nushell/pull/513

**Comment by wycats** [comment]
I reviewed this in more detail. I think this logic belongs outside of FileSystemShell, so that it would work regardless of which kind of Shell you were inside of.

```
$ enter someobject
$ cd -
```

^ should work

**Comment by twe4ked** [comment]
@wycats You're right. I'll update to support that. Hopefully I'll have time this weekend.

There is also [a bug](https://github.com/nushell/nushell/pull/363#discussion_r318628163) that @vivainio pointed out that I'll include a fix for. And it looks like there are conflicting files again that I need to fix!

I am still planning on finishing this!

**Comment by androbtech** [comment]
@twe4ked how are you doing? we've been wondering about you!

**Comment by twe4ked** [comment]
Thanks for reaching out @androbtech, I've just been really busy.

I've force pushed a new commit (there were conflicts again and I don't think the history was interesting) that passes on Windows and fixes the `cd foo/-` bug.

@wycats I couldn't see an obvious place to store the data at a higher level. It feels to me like it belongs on the shell structs. The implementation is going to need to be different for each shell anyway. We could make `ShellManager` `set_path()` update `last_path` but that would require implementing a `set_last_path()` which doesn't feel like it really adds anything. LMK if you have other ideas.

I haven't implemented the functionality for the value or help shells but I'm not sure that needs to block the PR, depending on the answer to the previous paragraph. I don't see any tests for the value/help shells either so I wasn't sure what the testing plan is.

**Pull request merged**